py/server/tests/test_udf_numpy_args.py

```diff
@@ -0,0 +1,384 @@
+    if np.isnan(col):
+        return True
+    else:
+        if not isinstance(col, {np_type}):
+            return True
+        return False
+"""
+                exec(func, globals())
+                with self.subTest(data_type):
+                    tbl = empty_table(100).update([col1_formula, col2_formula])
+                    # for floating point types, DH nulls are auto converted to np.nan
+                    # for integer types, DH nulls in the array raise exceptions
+                    if data_type in ("float", "double"):
+                        res = tbl.update("Col3 = test_udf(Col2)")
+                        self.assertEqual(4, res.to_string().count("true"))
+                    else:
+                        with self.assertRaises(DHError) as cm:
+                            res = tbl.update("Col3 = test_udf(Col2)")
+
+                func = f"""
+def test_udf(col: Optional[{np_type}]) -> bool:
+    if col is None:
+        return True
+    else:
+        if not isinstance(col, {np_type}):
+            return True
+        return False
+"""
+                exec(func, globals())
+                with self.subTest(data_type):
+                    tbl = empty_table(100).update([col1_formula, col2_formula])
+                    res = tbl.update("Col3 = test_udf(Col2)")
+                    self.assertEqual(4, res.to_string().count("true"))
+
+    def test_weird_cases(self):
+        def f(p1: Union[np.ndarray[typing.Any], None]) -> bool:
+            return bool(p1)
+
+        with self.assertRaises(DHError) as cm:
+            t = empty_table(10).update(["X1 = f(i)"])
+
+        def f1(p1: Union[np.int16, np.int32]) -> bool:
+            return bool(p1)
+
+        with self.assertRaises(DHError) as cm:
+            t = empty_table(10).update(["X1 = f1(i)"])
+
+        def f11(p1: Union[float, np.float32]) -> bool:
+            return bool(p1)
+
+        with self.assertRaises(DHError) as cm:
+            t = empty_table(10).update(["X1 = f11(i)"])
+
+        def f2(p1: Union[np.int16, np.float64]) -> Union[Optional[bool]]:
+            return bool(p1)
+
+        t = empty_table(10).update(["X1 = f2(i)"])
+        self.assertEqual(t.columns[0].data_type, dtypes.bool_)
+        self.assertEqual(9, t.to_string().count("true"))
+
+        def f21(p1: Union[np.int16, np.float64]) -> Union[Optional[bool], int]:
+            return bool(p1)
+
+        with self.assertRaises(DHError) as cm:
+            t = empty_table(10).update(["X1 = f21(i)"])
+
+        def f3(p1: Union[np.int16, np.float64], p2=None) -> bool:
+            return bool(p1)
+
+        t = empty_table(10).update(["X1 = f3(i)"])
+        self.assertEqual(t.columns[0].data_type, dtypes.bool_)
+
+        def f4(p1: Union[np.int16, np.float64], p2=None) -> bool:
+            return bool(p1)
+
+        t = empty_table(10).update(["X1 = f4((double)i)"])
+        self.assertEqual(t.columns[0].data_type, dtypes.bool_)
+        with self.assertRaises(DHError) as cm:
```

Comment:
It would be good if the error assumptions checked the messages to make sure we are failing the way we expect.